sintr_analysis_server_example/lib/query/perf.dart

```diff
@@ -1,71 +1,220 @@
+  sdkResults.putIfAbsent(perfName, () => 0);
+  ++sdkResults[perfName];
 
-final extractPerf = (String ln) {
-  _sanityCheck(ln);
-  List<String> splits = ln.split(":");
-  String msgType = splits[1];
+  return results;
+};
 
-  if (msgType == "Perf") return ln;
-  else return null;
+/// Merge two sets of results
+final apiUsageReductionMerge = perfReductionMerge;
 
-  int time = int.parse(splits[0].substring(1));
-  return [time, msgType];
-};
+/// Add an extraction result to the overall performance [results]
+/// where [extracted] is provided by [PerfMapper].
+final perfReducer = (String sdkVersion, List perfData, Map results) {
+  // Extract perf data
+  var perfType = perfData[2];
+  var perfName = perfData[3];
+  var elapsedTime = perfData[4];
 
-Map _reqMap = {};
-final extractReqResPerf = (String ln) {
-  _sanityCheck(ln);
-  List<String> splits = ln.split(":");
-  String msgType = splits[1];
+  // Sanity check
+  if (elapsedTime is! int) return results;
 
-  if (msgType != "Req" && msgType != "Res") return null;
-  int time = int.parse(splits[0].substring(1));
+  // Extract current results
+  Map sdkResults = results.putIfAbsent(sdkVersion, () => {});
+  Map perfTypeResults = sdkResults.putIfAbsent(perfType, () => {});
+  Map perfResults = perfTypeResults.putIfAbsent(perfName, () => {});
 
-//  ~1422652636123:Req:{"id"::"89","method"::"completion.getSuggestions","params"::{"file"::"...","offset"::152868},"clientRequestTime"::1422652634934}
-// ~1422652636123:Res:{"id"::"89","result"::{"id"::"11"}}
+  // Update current results
+  _updateBucket(perfResults, elapsedTime, [0, 1, 5, 50]);
+  return results;
+};
 
-  String dataArea = ln.substring("~1422652636123:Req:".length);
-  var dataMap = JSON.decode(dataArea.replaceAll("::", ":"));
-  var id = dataMap["id"];
+/// Merge two sets of results
+final perfReductionMerge = (Map results1, Map results2) {
+  Map results = {};
+  results1.forEach((key, value1) {
+    var value2 = results2[key];
+    if (value2 == null) {
+      results[key] = value1;
+    } else if (value1 is Map) {
+      results[key] = perfReductionMerge(value1, value2);
+    } else {
+      results[key] = value1 + value2;
+    }
+  });
+  results2.forEach((key, value2) {
+    var value1 = results1[key];
+    if (value1 == null) {
+      results[key] = value2;
+    }
+  });
+  return results;
+};
+
+/// Increment the count in the bucket containing [value]
+/// where [limits] are the bounds used for the initial set of buckets.
+/// Any values beyond the last bound specified in [limits]
+/// are placed into buckets of size increasing by a multiple of 2
+/// times the last bucket bounds.
+void _updateBucket(Map<int, int> buckets, int value, List<int> limits) {
+  var limitIter = limits.iterator..moveNext();
+  int limit = limitIter.current;
+  int lastIndex = 0;
+  while (limit < value) {
+    limit = limitIter.moveNext() ? limitIter.current : limit * 2;
+  }
+  buckets.putIfAbsent(limit, () => 0);
```

Comment:
You need to add the intermediary buckets as well, even if they're empty. Otherwise if you consider the data [0, 1, 2, 100], with buckets [0, 1, 2]

The buckets will end up looking like

[0, 1, 2, 128]

That's a bit unfortunate, as it requires the visualiser to have knowledge of the bucket generation algorithm to know that, for example, an item of value 5 would have got its own bucket rather than being lumped into the bucket '2'.

This can be solved by just moving the putIfAbsent call inside the while loop.